Sources/SwiftParser/Declarations.swift

```diff
@@ -311,6 +325,185 @@ extension Parser {
     )
   }
 
+  /// Parse the standalone header of a declaration group (a header that won't have a member block attached to it).
+  /// For a header that will have a member list attached, use ``Parser/parseHeaderForDeclarationGroup(attrs:keyword:handle:)``, which this method wraps.
+  ///
+  /// If `inMemberDeclList` is `true`, we know that the next item must be a
+  /// declaration and thus start with a keyword. This allows further recovery.
+  mutating func parseDeclarationGroupHeader(inMemberDeclList: Bool = false) -> RawDeclGroupHeaderSyntax {
+    // If we are at a `#if` of attributes, the `#if` directive should be
+    // parsed when we're parsing the attributes.
+    if self.at(.poundIf) && !self.withLookahead({ $0.consumeIfConfigOfAttributes() }) {
+      // If we're at a #if, that's not actually allowed here; parse it and add
+      // it as an unexpected node of a MissingDeclGroupHeaderSyntax.
```

Comment:
Why is this needed?

Reply:
If somebody—or some tool—mistakenly writes:
```
@abi(
    #if SomeCond
    class C1
    #endif
)
class C1 { ... }
```

I want the parser to recover and emit a good error.

I especially care about this because at one point, the compiler was emitting `#if $LangFeature` checks inside `@abi` attributes in module interfaces. I've fixed that by having the AST printer treat a usage inside an `@abi` attribute as a usage in the decl the `@abi` attribute is attached to, but I want to make sure that if there's a similar bug in the future, we parse it in a way that will produce a good diagnostic.

Reply:
Perhaps I should add a little commentary:

I've noticed that a weakness of the `@abi` syntax is that the parser struggles to recover when its syntax is malformed. I *think* the problem is that the parser's recovery system assumes that any parentheses it finds are part of the declaration, rather than part of an enclosing syntactic construct (the `@abi` attribute). I've therefore had to add some lookaheads and special cases which might not otherwise have been necessary.

If I'm holding it wrong or there is some way to modify the recovery system to work more gracefully for `@abi`, I'd love to hear it—I haven't worked much with this parser yet.

Reply:
I see about the `#if`. It makes sense with your explanation that we might incorrectly generate `#if` inside `@abi`. Otherwise, you could equally argue that the developer just forgot to finish the `@abi` attribute here:

```swift
@abi(

#if blah
class Foo {}
#endif
```

---

Do you have any concrete examples here? The problem is probably that `)` has a relatively low recovery precedence and decl keyword have a pretty high recovery precedence. Depends on the examples whether something can be done here.